src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets

```diff
@@ -55,20 +55,20 @@
     <Error Condition="'$(PublishAot)' == 'true' and '$(PublishSingleFile)' == 'true'" Text="PublishAot and PublishSingleFile cannot be specified at the same time." />
 
     <!-- Fail with descriptive error message for common unsupported cases. -->
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' == 'Windows_NT' and !$(RuntimeIdentifier.StartsWith('win'))"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' == 'Windows_NT' and '$(_targetOS)' != 'win'"
       Text="Cross-OS native compilation is not supported." />
 
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' != 'Windows_NT' and $(RuntimeIdentifier.StartsWith('win'))"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' != 'Windows_NT' and '$(_targetOS)' == 'win'"
       Text="Cross-OS native compilation is not supported." />
 
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and !($(RuntimeIdentifier.EndsWith('x64')) or $(RuntimeIdentifier.EndsWith('arm64')))"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(_targetArchitecture)' != 'x64' and '$(_targetArchitecture)' != 'arm64'"
       Text="Native compilation does not support targeting $(RuntimeIdentifier) yet." />
 
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and !('$(IlcHostArch)' == 'x64' or '$(IlcHostArch)' == 'arm64')"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and !('$(_hostArchitecture)' == 'x64' or '$(_hostArchitecture)' == 'arm64')"
       Text="Native compilation can run on x64 and arm64 hosts only." />
 
-    <Error Condition="'$(IlcHostPackagePath)' == '' and '$(RuntimePackagePath)' != '' and ('$(IlcHostArch)' == 'x64' or '$(IlcHostArch)' == 'arm64')"
-      Text="Add a PackageReference for '$(IlcHostPackageName)' to allow cross-compilation for $(_targetArchitecture)" />
+    <Error Condition="'$(_hostPackageName)' == '' and '$(RuntimePackagePath)' != '' and ('$(_hostArchitecture)' == 'x64' or '$(_hostArchitecture)' == 'arm64')"
+      Text="Add a PackageReference for '$(_hostPackageName)' to allow cross-compilation for $(_targetArchitecture)" />
```

Comment:
FWIW I don't think this will ever trigger correctly for SDK builds. It is a preexisting bug though.

Reply:
Better to delete it then?

Reply:
I think it could be deleted.

The only case where `_hostPackageName` could be empty is if there's some logic error, `RuntimeIdentifier` is not set, or `NETCoreSdkPortableRuntimeIdentifier` is not set. This target should only be reached if `RuntimeIdentifier` is set, so that leaves `NETCoreSdkPortableRuntimeIdentifier`. Can it even happen that `NETCoreSdkPortableRuntimeIdentifier` is not set (eg. for source builds or live build in CLR tree)? If it cannot happen then the checks for `'$(NETCoreSdkPortableRuntimeIdentifier)' != ''` are useless. If it can happen then the whole thing will likely fall apart with some hard to diagnose error.

It would be nice to have `'$(RuntimePackagePath)' == ''` check though to diagnose errors like #82542. The missing path could turn into hard to diagnose errors. For example, instead of collecting files from `$(RuntimePackagePath)/sdk/` it would try to collect files from `/sdk/` absolute path. That path could actually exist on the machine and be non-empty.

Reply:
Thanks for the catch. `IlcHostPackagePath` wasn't supposed to be renamed in the condition.

> If it cannot happen then the checks for '$(NETCoreSdkPortableRuntimeIdentifier)' != '' are useless. 

I think the `RuntimeIdentifier != ''` and `NETCoreSdkPortableRuntimeIdentifier != ''` checks are to prevent it from running on non-publish targets (dotnet build vs dotnet publish).

Reply:
`NETCoreSdkPortableRuntimeIdentifier` is set regardless of target (build vs. publish) as long as you use `Sdk="Microsoft.NET.Sdk"` (or other SDK that includes it transitively).

`RuntimeIdentifier` is set either with explicit property, through command line `-r` switch, or implicitly inferred when `dotnet publish` is used. It can be present both for build and publish but it's guaranteed to be present in `dotnet publish` (at least for recent SDKs that support PublishAot).

Reply:
Removed the redundant condition.

Reply:
> Removed the redundant condition.

Reverting this because CI started failing:

> artifacts/bin/coreclr/linux.x64.Debug/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(5,5): error MSB4184: The expression """.SubString(0, -1)" cannot be evaluated. 'length' must be a non-negative value. (Parameter 'length')

`NETCoreSdkPortableRuntimeIdentifier` is available during publish, but it may not be available for other targets restore/build etc.

Reply:
I'll check what is going on in that configuration. I did run empty MSBuild SDK project with single custom target to verify that it set `NETCoreSdkPortableRuntimeIdentifier` regardless of the target, so this must be some odd case.

Reply:
It failed while building `/t:TestBuild` target in `src/tests/build.proj`:

```
Building tests via "/Users/runner/work/1/s/eng/common/msbuild.sh" --ci --warnAsError false /Users/runner/work/1/s/src/tests/build.proj /t:TestBuild /p:TargetArchitecture=x64 /p:Configuration=Release /p:TargetOS=osx /nodeReuse:false    /maxcpucount "/flp:Verbosity=normal;LogFile=/Users/runner/work/1/s/artifacts/log/TestBuild.osx.x64.Release.log" "/flp1:WarningsOnly;LogFile=/Users/runner/work/1/s/artifacts/log/TestBuild.osx.x64.Release.wrn" "/flp2:ErrorsOnly;LogFile=/Users/runner/work/1/s/artifacts/log/TestBuild.osx.x64.Release.err" "/bl:/Users/runner/work/1/s/artifacts/log/TestBuild.osx.x64.Release.binlog" /p:NUMBER_OF_PROCESSORS=4 /p:LibrariesConfiguration=Release
/Users/runner/work/1/s/artifacts/bin/coreclr/osx.x64.Release/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(5,5): error MSB4184: The expression """.SubString(0, -1)" cannot be evaluated. 'length' must be a non-negative value. (Parameter 'length') [/Users/runner/work/1/s/src/tests/build.proj]
/Users/runner/work/1/s/artifacts/bin/coreclr/osx.x64.Release/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(5,5): error MSB4184: Actual value was -1. [/Users/runner/work/1/s/src/tests/build.proj]
##[error]artifacts/bin/coreclr/osx.x64.Release/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(5,5): error MSB4184: The expression """.SubString(0, -1)" cannot be evaluated. 'length' must be a non-negative value. (Parameter 'length')
Actual value was -1.
```

Reply:
That project uses the Traversal SDK and not Microsoft.NET.Sdk IIRC. I am not sure if it needs to reference ILCompiler build files. It's not the only odd thing in the test infrastructure...

Reply:
For posterity,
- `src/tests/build.proj` doesn't use `Microsoft.NET.Sdk`
- `src/tests/Directory.targets` imports `Microsoft.DotNet.ILCompiler.SingleEntry.targets` but there's no condition which would check it's actually `Microsoft.NET.Sdk` project build. There's many other live-build properties that make sense only for SDK-based projects.

I think it's okay to keep the extra check in `Microsoft.DotNet.ILCompiler.SingleEntry.targets`. However, it would potentially make sense to clean up `src/tests/Directory.targets`. We could split off the live-build properties into separate .targets file and include that one with `'$(NETCoreSdkPortableRuntimeIdentifier)' != ''` guard (or some more descriptive one). That way it won't import unnecessary stuff into the traversal project(s).

Reply:
I filed #82677 to track some of the issues with src/tests/* project files.